src/main/java/edu/unc/lib/boxc/migration/cdm/services/DestinationsService.java

```diff
@@ -0,0 +1,160 @@
+import java.util.List;
+import java.util.regex.Pattern;
+
+import org.apache.commons.csv.CSVFormat;
+import org.apache.commons.csv.CSVParser;
+import org.apache.commons.csv.CSVPrinter;
+import org.apache.commons.csv.CSVRecord;
+import org.slf4j.Logger;
+
+import edu.unc.lib.boxc.migration.cdm.exceptions.InvalidProjectStateException;
+import edu.unc.lib.boxc.migration.cdm.exceptions.MigrationException;
+import edu.unc.lib.boxc.migration.cdm.exceptions.StateAlreadyExistsException;
+import edu.unc.lib.boxc.migration.cdm.model.DestinationsInfo;
+import edu.unc.lib.boxc.migration.cdm.model.DestinationsInfo.DestinationMapping;
+import edu.unc.lib.boxc.migration.cdm.model.MigrationProject;
+import edu.unc.lib.boxc.migration.cdm.options.GenerateDestinationMappingOptions;
+import edu.unc.lib.boxc.migration.cdm.util.ProjectPropertiesSerialization;
+import edu.unc.lib.boxc.model.api.ids.RepositoryPathConstants;
+
+/**
+ * Service for working with destination mappings
+ * @author bbpennel
+ */
+public class DestinationsService {
+    private static final Logger log = getLogger(DestinationsService.class);
+
+    public static final String DESTINATION_FIELD = "boxc_dest";
+    public static final String COLLECTION_FIELD = "new_collection";
+    public static final String ID_FIELD = "id";
+    public static final String[] CSV_HEADERS = new String[] {
+            ID_FIELD, DESTINATION_FIELD, COLLECTION_FIELD };
+
+    public static final Pattern DEST_PATTERN = Pattern.compile(RepositoryPathConstants.UUID_PATTERN);
+
+    private MigrationProject project;
+
+    /**
+     * Generate a destination mapping file for the project using the provided options
+     * @param options
+     */
+    public void generateMapping(GenerateDestinationMappingOptions options) throws IOException {
+        assertProjectStateValid();
+        assertValidDestination(options.getDefaultDestination());
+        ensureMappingState(options.isForce());
+
+        Path fieldsPath = project.getDestinationMappingsPath();
+
+        try (
+            BufferedWriter writer = Files.newBufferedWriter(fieldsPath);
+            CSVPrinter csvPrinter = new CSVPrinter(writer, CSVFormat.DEFAULT.withHeader(CSV_HEADERS));
+        ) {
+            if (options.getDefaultDestination() != null) {
+                csvPrinter.printRecord(DestinationsInfo.DEFAULT_ID,
+                        options.getDefaultDestination(),
+                        options.getDefaultCollection());
+            }
+        }
+
+        project.getProjectProperties().setDestinationsGeneratedDate(Instant.now());
+        ProjectPropertiesSerialization.write(project);
+    }
+
+    private void assertProjectStateValid() {
+        if (project.getProjectProperties().getIndexedDate() == null) {
+            throw new InvalidProjectStateException("Project must be indexed prior to generating destinations");
+        }
+    }
+
+    private void assertValidDestination(String destination) {
+        if (destination == null) {
```

Comment:
Should this case throw an exception too?

Reply:
Updated